features/nvstore/source/nvstore.cpp

```diff
@@ -658,18 +645,20 @@ int NVStore::do_set(uint16_t key, uint16_t buf_len_bytes, const uint32_t *buf, u
         }
         else {
             // In the case we have crossed the limit, and the initial offset was also after the limit,
-            // this means we are not first writer (uncommon case). Just wait for GC to complete.
-            // then write record.
-            _write_lock.exclusive_unlock();
+            // this means we are not the first writer (uncommon case). Just wait for GC to complete.
+            // then write the record.
+            _write_lock.shared_unlock();
             for (;;) {
+                rtos::Thread::wait(MEDITATE_TIME_MS);
                 _write_lock.shared_lock();
-                if (save_active_area != _active_area) {
+                new_free_space = safe_increment(_free_space_offset, record_size);
+                // Use the same logics as before to check whether GC is complete
+                if (new_free_space < _size) {
```

Comment:
Still racy - while you were meditating lots of other people could have used the new area, and you might now turn out to be the person who is expected to actually do the GC on that area! (record_offset < _size). I still think goto 634 or its logical equivalent is required.

If you really have a philosophical objection to `goto`, then put the entire block in a `do {} while (0 or something)` and call it `continue` instead. But I think `goto retry` is fine.

Reply:
Thanks - embraced your solution (got no problems with goto - use it a lot in my code whenever needed).